Source/Core/Common/File.h

```diff
@@ -0,0 +1,234 @@
+  {
+    return WriteArray(reinterpret_cast<const char*>(data), length);
+  }
+
+  bool IsOpen() const override { return nullptr != m_file; }
+  // IsGood returns false after a read, write or other function fails
+  bool IsGood() const override { return m_good; }
+  // Reset IsGood to true
+  void Clear() override
+  {
+    m_good = true;
+    std::clearerr(m_file);
+  }
+
+  std::FILE* GetHandle() { return m_file; }
+  void SetHandle(std::FILE* file);
+
+  bool Seek(s64 off, int origin) override;
+  u64 Tell() const override;
+  u64 GetSize() override;
+  bool Resize(u64 size);
+  bool Flush();
+
+protected:
+  bool Read(void* data, size_t element_size, size_t count, size_t* pReadElements) override;
+
+private:
+  bool Write(const void* data, size_t element_size, size_t count);
+
+  std::FILE* m_file;
+  bool m_good;
+};
+
+class ReadOnlyFileBuffer : public std::streambuf
+{
+public:
+  ReadOnlyFileBuffer(ReadOnlyFile* file, size_t buffer_size);
+  int underflow() override;
+
+private:
+  ReadOnlyFile* m_file;
+  std::vector<char> m_buffer;
+};
+
+struct ReadOnlyFileStreamBase
+{
+  ReadOnlyFileBuffer m_buffer;
+
+  ReadOnlyFileStreamBase(ReadOnlyFile* file, size_t buffer_size) : m_buffer(file, buffer_size) {}
+};
+
+class ReadOnlyFileStream final : virtual ReadOnlyFileStreamBase, public std::istream
+{
+public:
+  // This object doesn't take ownership of the file object that's passed in.
+  // The file object must remain in memory while this object is being used.
+  ReadOnlyFileStream(ReadOnlyFile* file, size_t buffer_size = 65536)
+      : ReadOnlyFileStreamBase(file, buffer_size), std::ios(&m_buffer), std::istream(&m_buffer)
+  {
+    if (!*file)
+      setstate(std::ios::badbit);
+  }
+};
+
+class DirectoryIterator
+{
+public:
+  virtual ~DirectoryIterator() {}
+  // Advances this iterator to the next child and returns its name.
+  // An empty string is returned when the end is reached.
+  virtual std::string NextChild() = 0;
+};
+
+class StandardDirectoryIterator final : public DirectoryIterator, public NonCopyable
+{
+public:
+  StandardDirectoryIterator(const std::string& path);
+  ~StandardDirectoryIterator();
+
+  std::string NextChild() override;
+
+private:
+#ifdef _WIN32
+  HANDLE m_handle;
+  WIN32_FIND_DATA m_find_data;
+  bool m_first_child = true;
+#else
+  DIR* m_dir;
+#endif
+};
+
+class Path final
+{
+  friend class const_iterator;
+
+public:
+  class const_iterator final : public std::iterator<std::input_iterator_tag, Path>
+  {
+  public:
+    const_iterator(const Path& path);
+
+    const_iterator& operator++();
+
+    const Path& operator*() const { return *m_child_path; }
+    const Path* operator->() const { return m_child_path.get(); }
+    bool operator==(const const_iterator& other) const
+    {
+      return *m_child_path == *other.m_child_path;
+    }
+    bool operator!=(const const_iterator& other) const { return !(*this == other); }
+  private:
+    std::string m_base_path;
```

Comment:
This seems like a design smell, considering it's clearly doing more than just iterating.